Tool for scouting exportability in one shot

SherlockNoMad opened the pull request:
Summary:
Tool for scouting exportability issues in one shot. 

- Collect sample inputs for all submodules by running eager inference with forward_pre_hook. 
- Start from root module, recursively try exporting child modules, if current module export fails. 

Limitations: 
- only works for nn.module that contains tree-like submodules structure. this doesn't work for flatten GraphModule. 

TODO: support dynamic_dims


Sample output: https://docs.google.com/spreadsheets/d/1jnixrqBTYbWO_y6AaKA13XqOZmeB1MQAMuWL30dGoOg/edit?usp=sharing 

```
exportability_report = 
        {
            '': UnsupportedOperatorException(func=<OpOverload(op='testlib.op_missing_meta', overload='default')>),
            'submod_1': UnsupportedOperatorException(func=<OpOverload(op='testlib.op_missing_meta', overload='default')>),
            'submod_2': None
        }
```

Test Plan: buck2 run mode/dev-nosan fbcode//caffe2/test:test_export -- -r TestExportTools

Differential Revision: D57466486

zhxchen17 commented on torch/_export/tools.py:
```diff
@@ -0,0 +1,125 @@
+
+    kwargs = kwargs or {}
+
+    all_submod_names = [name for name, _ in mod.named_modules() if name != ""]
```
```suggestion
    log_export_usage(event="export.report_exportability")
    all_submod_names = [name for name, _ in mod.named_modules() if name != ""]
```

zhxchen17 reviewed (approved):
lint?

SherlockNoMad committed:
Tool for reporting exportability issues in one shot (#126471)

Summary:

Tool for report exportability issues in one shot. 

- Collect sample inputs for all submodules by running eager inference with forward_pre_hook. 
- Start from root module, recursively try exporting child modules, if current module export fails. 

Limitations: 
- only works for nn.module that contains tree-like submodules structure. this doesn't work for flatten GraphModule. 

TODO: support dynamic_dims


Sample output: https://docs.google.com/spreadsheets/d/1jnixrqBTYbWO_y6AaKA13XqOZmeB1MQAMuWL30dGoOg/edit?usp=sharing 

```
exportability_report = 
        {
            '': UnsupportedOperatorException(func=<OpOverload(op='testlib.op_missing_meta', overload='default')>),
            'submod_1': UnsupportedOperatorException(func=<OpOverload(op='testlib.op_missing_meta', overload='default')>),
            'submod_2': None
        }
```

Test Plan: buck2 run mode/dev-nosan fbcode//caffe2/test:test_export -- -r TestExportTools

Reviewed By: zimin2000, henryoier, zhxchen17

Differential Revision: D57466486
